server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java

```diff
@@ -75,14 +77,23 @@ public class KillUnusedSegments implements CoordinatorDuty
   private final Duration durationToRetain;
   private final boolean ignoreDurationToRetain;
   private final int maxSegmentsToKill;
+  private final Duration bufferPeriod;
 
   /**
    * Used to keep track of the last interval end time that was killed for each
    * datasource.
    */
   private final Map<String, DateTime> datasourceToLastKillIntervalEnd;
+
+  /**
+   * State that is maintained in the duty to determine if the duty needs to run or not.
+   */
   private DateTime lastKillTime;
-  private final Duration bufferPeriod;
+
+  /**
+   * Round-robin iterator of the datasources to kill. It's updated in every run by the duty.
+   */
+  private final RoundRobinIterator datasourceIterator;
```

Comment:
Is this iterator thread safe? What protects this duty from only having one thread running at a given time?

Reply:
Good point, the implementation is not thread safe. I've added the `@NotThreadSafe` annotation to this class.